src/app/static/src/app/components/genericChart/GenericChartTable.vue

```diff
@@ -72,30 +69,80 @@
             columnsWithHierarchy() {
                 return this.tableConfig.columns.filter(column => column.data.hierarchyColumn);
             },
+            hierarchyColumnPaths() {
+               // Generate a dictionary of dictionaries allowing hierarchy label lookup for all  hierarchy columns (i.e. area)
+               // to save the hierarchy string for each option. This traverses the entire hierarchy once only, which is
+               // quicker than calling findPath on each row value.
+               const addPathsToDict = (option: NestedFilterOption, parentPath: string,  dict: Dict<string>) => {
+                   //Add the path for this item...
+                   dict[option.id] = parentPath;
+                   //...then adds paths for any children
+                   const path = parentPath + (parentPath.length ? "/" : "") + option.label;
+                   if (option.children) {
+                       option.children.forEach((child) => addPathsToDict(child as NestedFilterOption, path, dict));
+                   }
+               };
+
+               const result = {} as Dict<Dict<string>>;
+               this.columnsWithHierarchy.forEach((columnConfig) => {
+                   const columnId = columnConfig.data.hierarchyColumn!;
+                   const pathsDict = {};
+                   const column = this.columns.find((column) => column.id == columnId)!;
+                   // Deal with both array or object as top-level values type
+                   const values = Array.isArray(column.values) ? column.values : [column.values];
+                   values.forEach((value) => addPathsToDict(value, "", pathsDict));
+                   result[columnId] = pathsDict;
+               });
+               return result;
+            },
             labelledData() {
                 // Build a dictionary of the columns defined in the fetched dataset, which include all values and
                 // associated labels
                 const columnsDict = this.columns.reduce((dict, column) => ({...dict, [column.id]: column}), {} as Dict<GenericChartColumn>);
-                return this.filteredData.map(row => {
+
+                // Convert each label column's filter options to a dictionary of id to labels so we do not need to find values'
+                // labels in the options array per row
+                const configLabelColumns = this.tableConfig.columns.filter(column => column.data.labelColumn);
+                const columnValueLabels = configLabelColumns.reduce((dict, columnConfig) => {
```

Comment:
For readability, can we have reduce function separated in such a way that it makes the code easy to read.

`function reducer(previousValue, currentValue) { doSomething }`
`configLabelColumns.reduce(reducer)`